File: dynabuffers-python/dynabuffers/NamespaceResolver.py
from math import ceil
from typing import List, Union, Any

from dynabuffers.api.ISerializable import ISerializable, ByteBuffer
from dynabuffers.ast.NamespaceType import NamespaceType


class Waypoint:
    def __init__(self, name: str, position: int):
        self._name = name
        self._position = position

    @property
    def name(self):
        return self._name

    @property
    def position(self):
        return self._position


class NamespaceDescription:
    def __init__(self, namespace: NamespaceType, path: List[Waypoint]):
        self._namespace = namespace
        self._path = path

    @property
    def namespace(self):
        return self._namespace

    @property
    def path(self):
        return self._path

    def joined_path(self) -> str:
        return ".".join(list(map(lambda waypoint: waypoint.name, self._path)))


class NamespaceResolver:
    def __init__(self, tree: List[ISerializable]):
        self.tree = tree

    def get_namespace_from_names(self, names: Union[List[str], None]) -> Union[NamespaceDescription, None]:
        if names is None:
            return self._infer_default_namespace()
        r = self.__get_namespace(names, self.__get_namespaces_from_root_level())
        return r

    def get_namespace(self, d: Union[dict, Any]) -> Union[NamespaceDescription, None]:
        namespaces = d.get(':namespace')
        if namespaces is not None:
            return self.get_namespace_from_names(namespaces.split('.'))
        return self._infer_default_namespace()

    def get_namespace_from_serialized(self, bb: ByteBuffer, len_namespace_path: int) -> Union[
        NamespaceDescription, None]:
        """
        Reads the encoded namespace path directly from the given byte buffer
        WARNING: Modifies the provided byte buffer
        :param bb: byte buffer of serialized payload
        :param len_namespace_path: denotes the length of the namespace path (in terms of namespaces)
        :return:
        """
        if len_namespace_path == 0:
            return None

        namespace_bytes = [bb.get() for _ in range(ceil(len_namespace_path / 2))]
        waypoints = []
        # Read waypoints from bytes, 2 waypoints per byte
        for byte in namespace_bytes:
            waypoints.append(byte & 0xF0)
            waypoints.append(byte & 0x0F)
        waypoints = waypoints[:len_namespace_path]

        # map the read path to the namespaces within the schema - sorry this is not very elegant
        namespaces = list(map(lambda nd: nd.namespace, self.__get_namespaces_from_root_level()))
        path = []
        for waypoint in waypoints:
            namespace = namespaces[waypoint]
            namespaces = namespace.nestedNamespaces
            path.append(NamespaceDescription(namespace, [Waypoint(namespace.options.name, waypoint)]))

        return NamespaceDescription(path[-1:][0].namespace, list(map(lambda n: n.path[0], path)))

    def __get_namespace(self, names: List[str], namespaces: List[NamespaceDescription]) -> Union[
        NamespaceDescription, None]:
        ns_name = names[0]
        ns = self.__find_namespace_by_name(ns_name, namespaces)
        if len(names) == 1:
            return ns

        nested = [NamespaceDescription(n, ns.path + [Waypoint(n.options.name, idx)]) for idx, n in
                  enumerate(ns.namespace.nestedNamespaces)]
        return self.__get_namespace(names[1:], nested)

    def __get_namespaces_from_root_level(self) -> List[NamespaceDescription]:
        namespace_types = list(filter(lambda x: isinstance(x, NamespaceType), self.tree))
        return [NamespaceDescription(ns, [Waypoint(ns.options.name, idx)]) for idx, ns in enumerate(namespace_types)]

    def __find_namespace_by_name(self, name: str, namespaces: [NamespaceDescription]) -> NamespaceDescription:
        namespace = next((x for x in namespaces if x.namespace.options.name == name), None)
        if namespace is None:
            raise ValueError("no namespace with name " + str(name) + " found")
        return namespace

    def _infer_default_namespace(self) -> Union[NamespaceDescription, None]:
        """
        infers the namespace from the schema - multiple nested namespaces will cause this method to raise an exception.
        :return:
        """

        def _resolve(namespaces: List[NamespaceDescription]) -> Union[NamespaceDescription, None]:
            if len(namespaces) == 0:
                return None
            if len(namespaces) > 1:
                raise ValueError("Could not infer default namespace")
            candidate = namespaces[0]
            if len(candidate.namespace.nestedNamespaces) == 0:
                return candidate
            new_candidates = [NamespaceDescription(ns, candidate.path + [Waypoint(ns.options.name, idx)]) for idx, ns in
                              enumerate(candidate.namespace.nestedNamespaces)]
            return _resolve(new_candidates)

        return _resolve(self.__get_namespaces_from_root_level())

File: dynabuffers-java/src/main/kotlin/dynabuffers/NamespaceResolver.kt
package dynabuffers

import dynabuffers.api.IType
import dynabuffers.ast.NamespaceType
import dynabuffers.exception.DynabuffersException
import java.nio.ByteBuffer
import kotlin.experimental.and
import kotlin.streams.toList

class NamespaceResolver(private val tree: List<IType>) {

    /**
     * Gets a namespace from the given list or tries to infer ([inferDefaultNamespace]) the namespace if no names are passed.
     */
    fun getNamespace(names: List<String>?): NamespaceDescription {
        if (names == null) {
            return inferDefaultNamespace()
        }
        return getNamespace(names, getNamespacesFromRootLevel())
    }

    /**
     * Extracts the namespace description from the given map. Looks for a [SpecialKey.NAMESPACE] key which
     * should contain the stringified path of the namespace within the tree.
     */
    fun getNamespace(map: Map<String, Any?>): NamespaceDescription {
        return getNamespace((map[SpecialKey.NAMESPACE.key] as String?)?.split("."))
    }

    /**
     * Extracts the namespace description out of a serialized dynabuffers message wrapped as a [ByteBuffer]
     * WARNING: this method modifies the given ByteBuffer
     */
    fun getNamespace(serialized: ByteBuffer, namespaceDepth: Int): NamespaceDescription {
        if (namespaceDepth == 0) return AbsentNamespace

        val namespacePathBuffer = ByteArray(kotlin.math.ceil(namespaceDepth / 2.0).toInt())
        serialized.get(namespacePathBuffer)
        val path = namespacePathBuffer.toList()
            .flatMap { listOf((it.toInt() shr 4).toByte(), it and 0x0F) }
            .take(namespaceDepth)

        return getNamepacesFromPath(path)
    }

    fun getNamepacesFromPath(path: List<Byte>): NamespaceDescription {
        val nss = getNamespaceFromPath(path)
        val waypoints = nss.map { it.path.first() }
        return ConcreteNamespaceDescription(checkNotNull(nss.lastOrNull()).namespace, waypoints)
    }

    /**
     * Reassembles the namespaces from the given path - A path is the byte sequence of the namespaces position
     * within the IDL file.
     *
     * eg:
     *
     *  namespace a {  // path => [0]
     *     namespace b { .. } // path => [0, 0]
     *     namespace c { .. } // path => [0, 1]
     *  }
     *
     *  and so forth.
     */
    private fun getNamespaceFromPath(path: List<Byte>): List<ConcreteNamespaceDescription> {
        val iter = path.iterator()
        var nss = getNamespacesFromRootLevel()
                .map { it.namespace }

        var ns: NamespaceType?
        var p = emptyList<ConcreteNamespaceDescription>()

        while (iter.hasNext()) {
            val i = iter.next()
            ns = nss[i.toInt()]
            nss = ns.nestedNamespaces
            p = p.plus(ConcreteNamespaceDescription(ns, listOf(Waypoint(ns.options.name, i))))
        }
        return p
    }

    /**
     * Tries to infer the namespace by traversing through the existing namespaces extracing the first one during each step.
     * If more than one namespace is encounted during a step a [IllegalStateException] is thrown since it is not possible
     * to infer a 'default' namespace.
     */
    private fun inferDefaultNamespace(): NamespaceDescription {
        tailrec fun _resolve(namespaces: List<ConcreteNamespaceDescription>): NamespaceDescription {
            if (namespaces.isEmpty()) return AbsentNamespace
            if (namespaces.size > 1) throw DynabuffersException("Could not infer default namespace")
            val candidate = namespaces.first()
            if (candidate.namespace.nestedNamespaces.isEmpty()) return candidate
            val newCandidates = candidate.namespace.nestedNamespaces
                    .mapIndexed { index, namespaceType ->
                        ConcreteNamespaceDescription(namespaceType,
                                candidate.path.plus(Waypoint(namespaceType.options.name, index.toByte())))
                    }
            return _resolve(newCandidates)
        }

        return _resolve(getNamespacesFromRootLevel())
    }

    private fun getNamespacesFromRootLevel(): List<ConcreteNamespaceDescription> =
            tree.filterIsInstance<NamespaceType>()
                    .mapIndexed { idx, namespace -> ConcreteNamespaceDescription(namespace, listOf(Waypoint(namespace.options.name, idx.toByte()))) }

    private fun getNamespace(names: List<String>, namespaces: List<ConcreteNamespaceDescription>): ConcreteNamespaceDescription {
        val nsName = names.first()
        val ns = findNamespaceByName(nsName, namespaces)
        if (names.size == 1) return ns
        return getNamespace(names.stream().skip(1).toList(),
                ns.namespace.nestedNamespaces
                        .mapIndexed { idx, n -> ConcreteNamespaceDescription(n, ns.path.plus(Waypoint(n.options.name, idx.toByte()))) }
        )
    }

    private fun findNamespaceByName(name: String, namespaces: List<ConcreteNamespaceDescription>): ConcreteNamespaceDescription {
        return namespaces.find { it.namespace.options.name == name }
                ?: throw DynabuffersException("no namespace with name $name found")
    }
}

sealed class NamespaceDescription() {
    abstract fun joinedPath(): String
    abstract fun getWaypoints(): List<Waypoint>
}

class ConcreteNamespaceDescription(val namespace: NamespaceType, val path: List<Waypoint>) : NamespaceDescription() {
    override fun joinedPath(): String = path.joinToString(".") { it.name }
    override fun getWaypoints(): List<Waypoint> {
        return path
    }

}

data class Waypoint(val name: String, val position: Byte)
object AbsentNamespace : NamespaceDescription() {
    override fun joinedPath(): String = throw IllegalStateException("No namespace has no name")
    override fun getWaypoints(): List<Waypoint> {
        return emptyList()
    }
}
